File: searchlib/src/main/java/com/yahoo/searchlib/aggregation/MaxAggregationResult.java
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.searchlib.aggregation;

import com.yahoo.searchlib.expression.ResultNode;
import com.yahoo.searchlib.expression.SingleResultNode;
import com.yahoo.vespa.objects.Deserializer;
import com.yahoo.vespa.objects.ObjectVisitor;
import com.yahoo.vespa.objects.Serializer;

/**
 * This is an aggregated result holding the maximum result of the matching hits.
 *
 * @author baldersheim
 * @author Simon Thoresen Hult
 */
public class MaxAggregationResult extends AggregationResult {

    public static final int classId = registerClass(0x4000 + 83, MaxAggregationResult.class);
    private SingleResultNode max;

    /**
     * Constructs an empty result node. <b>NOTE:</b> This instance is broken until non-optional member data is set.
     */
    public MaxAggregationResult() {

    }

    /**
     * Constructs an instance of this class with given max value.
     *
     * @param max The initial maximum to set.
     */
    public MaxAggregationResult(SingleResultNode max) {
        setMax(max);
    }

    /**
     * Returns the maximum value found in all matching hits.
     *
     * @return The value.
     */
    public final SingleResultNode getMax() {
        return max;
    }

    /**
     * Sets the maximum value found in all matching hits.
     *
     * @param max The value.
     * @return This, to allow chaining.
     */
    public final MaxAggregationResult setMax(SingleResultNode max) {
        this.max = max;
        return this;
    }

    @Override
    public ResultNode getRank() {
        return max;
    }

    @Override
    protected int onGetClassId() {
        return classId;
    }

    @Override
    protected void onSerialize(Serializer buf) {
        super.onSerialize(buf);
        serializeOptional(buf, max);
    }

    @Override
    protected void onDeserialize(Deserializer buf) {
        super.onDeserialize(buf);
        max = (SingleResultNode)deserializeOptional(buf);
    }

    @Override
    protected void onMerge(AggregationResult result) {
        max.max(((MaxAggregationResult)result).max);
    }

    @Override
    protected boolean equalsAggregation(AggregationResult obj) {
        return equals(max, ((MaxAggregationResult)obj).max);
    }

    @Override
    public MaxAggregationResult clone() {
        MaxAggregationResult obj = (MaxAggregationResult)super.clone();
        if (max != null) {
            obj.max = (SingleResultNode)max.clone();
        }
        return obj;
    }

    @Override
    public void visitMembers(ObjectVisitor visitor) {
        super.visitMembers(visitor);
        visitor.visit("max", max);
    }
}

File: searchlib/src/vespa/searchlib/aggregation/maxaggregationresult.cpp
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

#include "maxaggregationresult.h"

#include <vespa/document/fieldvalue/document.h>
#include <vespa/searchlib/expression/floatresultnode.h>
#include <vespa/searchlib/expression/resultvector.h>

#include <vespa/searchlib/aggregation/aggregation.hpp>
#include <vespa/vespalib/objects/visit.hpp>

#include <limits>

using search::expression::FloatResultNode;
using search::expression::NumericResultNode;
using search::expression::ResultNodeVector;
using vespalib::Deserializer;
using vespalib::Serializer;

namespace search::aggregation {

IMPLEMENT_AGGREGATIONRESULT(MaxAggregationResult, AggregationResult);

MaxAggregationResult::MaxAggregationResult()
    : AggregationResult(), _max(FloatResultNode(-std::numeric_limits<double>::max())) {
}

MaxAggregationResult::MaxAggregationResult(const SingleResultNode& max) : AggregationResult(), _max(max) {
}

MaxAggregationResult::~MaxAggregationResult() = default;

void MaxAggregationResult::onPrepare(const ResultNode& result) {
    if (!is_ready(_max.get(), result)) {
        _max = create_and_ensure_wanted<SingleResultNode, FloatResultNode>(result);
        _max->setMin();
    }
}

void MaxAggregationResult::initForUnitTest(const ResultNode& result) {
    if (!is_ready(_max.get(), result)) {
        _max = create_and_ensure_wanted<SingleResultNode, FloatResultNode>(result);
    }
    _max->set(result);
}

void MaxAggregationResult::onMerge(const AggregationResult& b) {
    _max->max(*static_cast<const MaxAggregationResult&>(b)._max);
}

void MaxAggregationResult::onAggregate(const ResultNode& result) {
    if (result.isMultiValue()) {
        static_cast<const ResultNodeVector&>(result).flattenMax(*_max);
    } else {
        _max->max(result);
    }
}

void MaxAggregationResult::onReset() {
    _max.reset(static_cast<SingleResultNode*>(_max->getClass().create()));
    _max->setMin();
}

Serializer& MaxAggregationResult::onSerialize(Serializer& os) const {
    AggregationResult::onSerialize(os);
    return os << _max;
}

Deserializer& MaxAggregationResult::onDeserialize(Deserializer& is) {
    AggregationResult::onDeserialize(is);
    return is >> _max;
}

void MaxAggregationResult::visitMembers(vespalib::ObjectVisitor& visitor) const {
    AggregationResult::visitMembers(visitor);
    visit(visitor, "max", _max);
}

} // namespace search::aggregation
